File: internal/innerchan/innerchan_test.go
package innerchan

import (
	"bytes"
	"crypto/ecdh"
	"crypto/ed25519"
	"crypto/rand"
	"encoding/binary"
	"encoding/hex"
	"errors"
	"sort"
	"sync"
	"testing"
)

// playClient mirrors what the browser does with a server handshake result.
func playClient(t *testing.T, clientEph *ecdh.PrivateKey, clientNonce []byte, res *HandshakeResult) *Session {
	t.Helper()
	th := TranscriptHash(clientEph.PublicKey().Bytes(), res.ServerEphPub, clientNonce, res.ServerEdPub)
	if !VerifyServerSignature(res.ServerEdPub, th, res.Sig) {
		t.Fatal("server signature does not verify")
	}
	peer, err := ecdh.X25519().NewPublicKey(res.ServerEphPub)
	if err != nil {
		t.Fatal(err)
	}
	ss, err := clientEph.ECDH(peer)
	if err != nil {
		t.Fatal(err)
	}
	sess, err := DeriveSession(ss, th)
	if err != nil {
		t.Fatal(err)
	}
	return sess
}

func TestHandshakeAndSealedFrames(t *testing.T) {
	b64, err := GenerateServerKey()
	if err != nil {
		t.Fatal(err)
	}
	priv, err := ParseServerKey(b64)
	if err != nil {
		t.Fatal(err)
	}
	clientEph, _ := ecdh.X25519().GenerateKey(rand.Reader)
	nonce := make([]byte, 32)
	rand.Read(nonce)

	res, err := ServerHandshake(priv, clientEph.PublicKey().Bytes(), nonce)
	if err != nil {
		t.Fatal(err)
	}
	client := playClient(t, clientEph, nonce, res)
	server := res.Session

	// server -> client, in order
	for i, msg := range []string{"welcome", "message one", "message two"} {
		f, err := server.SealToClient([]byte(msg))
		if err != nil {
			t.Fatal(err)
		}
		pt, err := client.OpenToClient(f)
		if err != nil || string(pt) != msg {
			t.Fatalf("frame %d: %q %v", i, pt, err)
		}
		// replay of the same frame is refused
		if _, err := client.OpenToClient(f); !errors.Is(err, ErrCounter) {
			t.Fatalf("replay: want ErrCounter, got %v", err)
		}
	}
	// client -> server, independent counter
	f1, _ := client.SealFromClient([]byte("hello"))
	f2, _ := client.SealFromClient([]byte("send"))
	if _, err := server.OpenFromClient(f2); !errors.Is(err, ErrCounter) {
		t.Fatalf("out of order: want ErrCounter, got %v", err)
	}
	if pt, err := server.OpenFromClient(f1); err != nil || string(pt) != "hello" {
		t.Fatalf("f1: %q %v", pt, err)
	}
	if pt, err := server.OpenFromClient(f2); err != nil || string(pt) != "send" {
		t.Fatalf("f2: %q %v", pt, err)
	}
	// tampered ciphertext with a correct counter
	f3, _ := client.SealFromClient([]byte("tamper me"))
	f3[len(f3)-1] ^= 1
	if _, err := server.OpenFromClient(f3); !errors.Is(err, ErrOpen) {
		t.Fatalf("tamper: want ErrOpen, got %v", err)
	}
	// direction confusion: a client frame is not a server frame
	f4, _ := client.SealFromClient([]byte("wrong way"))
	other := playClient(t, clientEph, nonce, res) // fresh counters
	if _, err := other.OpenToClient(f4); err == nil {
		t.Fatal("a c2s frame must not open as s2c")
	}
}

func TestSignatureBindsEverything(t *testing.T) {
	b64, _ := GenerateServerKey()
	priv, _ := ParseServerKey(b64)
	clientEph, _ := ecdh.X25519().GenerateKey(rand.Reader)
	nonce := make([]byte, 32)
	rand.Read(nonce)
	res, err := ServerHandshake(priv, clientEph.PublicKey().Bytes(), nonce)
	if err != nil {
		t.Fatal(err)
	}
	good := TranscriptHash(clientEph.PublicKey().Bytes(), res.ServerEphPub, nonce, res.ServerEdPub)
	if !VerifyServerSignature(res.ServerEdPub, good, res.Sig) {
		t.Fatal("good transcript must verify")
	}
	// any field substituted -> the signature no longer verifies
	otherEph, _ := ecdh.X25519().GenerateKey(rand.Reader)
	otherNonce := make([]byte, 32)
	rand.Read(otherNonce)
	otherPub, _, _ := ed25519.GenerateKey(rand.Reader)
	for name, th := range map[string][32]byte{
		"client-eph": TranscriptHash(otherEph.PublicKey().Bytes(), res.ServerEphPub, nonce, res.ServerEdPub),
		"server-eph": TranscriptHash(clientEph.PublicKey().Bytes(), otherEph.PublicKey().Bytes(), nonce, res.ServerEdPub),
		"nonce":      TranscriptHash(clientEph.PublicKey().Bytes(), res.ServerEphPub, otherNonce, res.ServerEdPub),
		"server-key": TranscriptHash(clientEph.PublicKey().Bytes(), res.ServerEphPub, nonce, otherPub),
	} {
		if VerifyServerSignature(res.ServerEdPub, th, res.Sig) {
			t.Fatalf("%s substituted: must not verify", name)
		}
	}
	// a MITM with its own key cannot re-sign the transcript against the pin
	if VerifyServerSignature(otherPub, good, res.Sig) {
		t.Fatal("signature must not verify under another key")
	}
	// bad sizes
	if VerifyServerSignature(res.ServerEdPub, good, res.Sig[:63]) {
		t.Fatal("63-byte signature must not verify")
	}
	if _, err := ParseServerKey("not base64!"); !errors.Is(err, ErrBadServerKey) {
		t.Fatalf("bad key: %v", err)
	}
	if _, err := ServerHandshake(priv, clientEph.PublicKey().Bytes()[:31], nonce); !errors.Is(err, ErrBadLength) {
		t.Fatalf("short eph: %v", err)
	}
}

// TestKnownAnswer freezes the derivation so the browser implementation
// (web/src/crypto/innerchan.ts) can assert byte-identical output from the
// same fixed inputs. Fixed X25519 scalars and a fixed server seed; the
// expected values were produced by this test and are asserted by the TS
// suite -- two implementations agreeing, not one copying the other.
func TestKnownAnswer(t *testing.T) {
	curve := ecdh.X25519()
	clientEph, err := curve.NewPrivateKey(bytes.Repeat([]byte{0x11}, 32))
	if err != nil {
		t.Fatal(err)
	}
	serverEph, err := curve.NewPrivateKey(bytes.Repeat([]byte{0x22}, 32))
	if err != nil {
		t.Fatal(err)
	}
	priv := ed25519.NewKeyFromSeed(bytes.Repeat([]byte{0x33}, 32))
	nonce := bytes.Repeat([]byte{0x44}, 32)
	serverEdPub := priv.Public().(ed25519.PublicKey)
	th := TranscriptHash(clientEph.PublicKey().Bytes(), serverEph.PublicKey().Bytes(), nonce, serverEdPub)
	ss, err := clientEph.ECDH(serverEph.PublicKey())
	if err != nil {
		t.Fatal(err)
	}
	sess, err := DeriveSession(ss, th)
	if err != nil {
		t.Fatal(err)
	}
	frame, err := sess.SealToClient([]byte("known answer"))
	if err != nil {
		t.Fatal(err)
	}
	sig := ed25519.Sign(priv, signedMessage(th))
	got := map[string]string{
		"transcript": hex.EncodeToString(th[:]),
		"frame":      hex.EncodeToString(frame),
		"sig":        hex.EncodeToString(sig),
	}
	want := map[string]string{
		"transcript": "5f33db1397e149e700d156a90fcc77da723e639da0b24023b863996b0effed20",
		"frame":      "0000000000000001a641c79613e5e8ea5dbcfde7486a901fd3daf5b943a36cce5aaa6d90",
		"sig":        "64f58b3b5cc056f6203a62400dc87b79a5c83e8c795b214bb5d8d330684b6dc42c65fdae068f5a06201f6520b1d5b81c71c8b7fdae810746881a6380dad0c30e",
	}
	for k, v := range want {
		if got[k] != v {
			t.Fatalf("KAT %s: got %s want %s", k, got[k], v)
		}
	}
	// The TS suite pins these; here we pin the structure that makes them
	// reproducible (counter 1, 8-byte prefix, 16-byte tag).
	if frame[7] != 1 || len(frame) != 8+len("known answer")+16 {
		t.Fatalf("frame shape: %x", frame)
	}
}

// Third audit: N goroutines sealing concurrently must mint frames that,
// ordered by their counters, all open -- the per-session mutex hands out
// counters exactly once each. (Ordering counter-to-WIRE is the writer
// mutex in ws.go's writeOne: seal+write are one critical section there.)
func TestConcurrentSealCountersAreDense(t *testing.T) {
	b64, _ := GenerateServerKey()
	priv, _ := ParseServerKey(b64)
	clientEph, _ := ecdh.X25519().GenerateKey(rand.Reader)
	nonce := make([]byte, 32)
	rand.Read(nonce)
	res, err := ServerHandshake(priv, clientEph.PublicKey().Bytes(), nonce)
	if err != nil {
		t.Fatal(err)
	}
	client := playClient(t, clientEph, nonce, res)

	const n = 64
	frames := make([][]byte, n)
	var wg sync.WaitGroup
	for i := 0; i < n; i++ {
		wg.Add(1)
		go func(i int) {
			defer wg.Done()
			f, err := res.Session.SealToClient([]byte{byte(i)})
			if err != nil {
				t.Error(err)
				return
			}
			frames[i] = f
		}(i)
	}
	wg.Wait()
	sort.Slice(frames, func(a, b int) bool {
		return binary.BigEndian.Uint64(frames[a][:8]) < binary.BigEndian.Uint64(frames[b][:8])
	})
	for i, f := range frames {
		if got := binary.BigEndian.Uint64(f[:8]); got != uint64(i+1) {
			t.Fatalf("counters not dense: frame %d has counter %d", i, got)
		}
		if _, err := client.OpenToClient(f); err != nil {
			t.Fatalf("frame with counter %d failed to open: %v", i+1, err)
		}
	}
}

File: web/src/crypto/innerchan.test.ts
// Tests for crypto/innerchan.ts -- 83-6. The known-answer test reproduces
// internal/innerchan's TestKnownAnswer from the same fixed inputs: the
// transcript hash must match, the Go-produced signature must verify, and
// the Go-sealed frame must open -- two implementations agreeing on every
// byte of the frozen construction.

import { test } from "node:test";
import { strict as assert } from "node:assert";

import { startClientHandshake, transcriptHash, deriveSession, serverFingerprint } from "./innerchan";

function hex(b: Uint8Array): string {
  return [...b].map((x) => x.toString(16).padStart(2, "0")).join("");
}
function unhex(s: string): Uint8Array {
  return new Uint8Array(s.match(/.{2}/g)!.map((h) => parseInt(h, 16)));
}
const fill = (b: number) => new Uint8Array(32).fill(b);

const PKCS8_ED25519 = unhex("302e020100300506032b657004220420");
async function ed25519PubFromSeed(seed: Uint8Array): Promise<{ pub: Uint8Array; priv: CryptoKey }> {
  const pkcs8 = new Uint8Array(PKCS8_ED25519.length + 32);
  pkcs8.set(PKCS8_ED25519, 0);
  pkcs8.set(seed, PKCS8_ED25519.length);
  const tmp = await crypto.subtle.importKey("pkcs8", pkcs8, { name: "Ed25519" }, true, ["sign"]);
  const jwk = await crypto.subtle.exportKey("jwk", tmp);
  const b64 = jwk.x!.replace(/-/g, "+").replace(/_/g, "/");
  const bin = atob(b64 + "=".repeat((4 - (b64.length % 4)) % 4));
  const pub = new Uint8Array([...bin].map((c) => c.charCodeAt(0)));
  const priv = await crypto.subtle.importKey("pkcs8", pkcs8, { name: "Ed25519" }, false, ["sign"]);
  return { pub, priv };
}

async function x25519PubFromScalar(scalar: Uint8Array): Promise<Uint8Array> {
  const hs = await startClientHandshake({ ephSeed: scalar, nonce: fill(0) });
  return hs.hello.clientEphPub;
}

// Values asserted by internal/innerchan/innerchan_test.go TestKnownAnswer.
const KAT = {
  transcript: "5f33db1397e149e700d156a90fcc77da723e639da0b24023b863996b0effed20",
  frame: "0000000000000001a641c79613e5e8ea5dbcfde7486a901fd3daf5b943a36cce5aaa6d90",
  sig: "64f58b3b5cc056f6203a62400dc87b79a5c83e8c795b214bb5d8d330684b6dc42c65fdae068f5a06201f6520b1d5b81c71c8b7fdae810746881a6380dad0c30e",
};

test("known answer: transcript, signature and a Go-sealed frame agree byte for byte", async () => {
  const serverEphPub = await x25519PubFromScalar(fill(0x22));
  const server = await ed25519PubFromSeed(fill(0x33));
  const hs = await startClientHandshake({ ephSeed: fill(0x11), nonce: fill(0x44) });
  const th = await transcriptHash(hs.hello.clientEphPub, serverEphPub, hs.hello.clientNonce, server.pub);
  assert.equal(hex(th), KAT.transcript);
  // finish() verifies the Go signature and derives the same keys
  const sess = await hs.finish(serverEphPub, server.pub, unhex(KAT.sig), server.pub);
  const pt = await sess.open(unhex(KAT.frame));
  assert.equal(new TextDecoder().decode(pt), "known answer");
});

test("finish refuses a bad signature, a substituted key, and a non-pinned key", async () => {
  const serverEphPub = await x25519PubFromScalar(fill(0x22));
  const server = await ed25519PubFromSeed(fill(0x33));
  const other = await ed25519PubFromSeed(fill(0x55));
  const hs = await startClientHandshake({ ephSeed: fill(0x11), nonce: fill(0x44) });
  const badSig = unhex(KAT.sig);
  badSig[10] ^= 1;
  await assert.rejects(hs.finish(serverEphPub, server.pub, badSig, null));
  // MITM presents its own key: the signature was made by the real server
  await assert.rejects(hs.finish(serverEphPub, other.pub, unhex(KAT.sig), null));
  // valid signature, but not the key we pinned -> the wall
  await assert.rejects(hs.finish(serverEphPub, server.pub, unhex(KAT.sig), other.pub), /not the pinned key/);
  // TOFU (no pin) accepts
  await hs.finish(serverEphPub, server.pub, unhex(KAT.sig), null);
});

test("counters: replay and reordering of inbound frames are refused; outbound increases", async () => {
  // Build two sessions from the same secret so we can seal "as the server"
  // using the mirror trick: a session's c2s key sealed by one side is the
  // other's s2c only if derived with swapped infos -- so instead exercise
  // the inbound path with the KAT frame and its own counter rules.
  const serverEphPub = await x25519PubFromScalar(fill(0x22));
  const server = await ed25519PubFromSeed(fill(0x33));
  const hs = await startClientHandshake({ ephSeed: fill(0x11), nonce: fill(0x44) });
  const sess = await hs.finish(serverEphPub, server.pub, unhex(KAT.sig), null);
  const frame = unhex(KAT.frame);
  await sess.open(frame);
  await assert.rejects(sess.open(frame), /counter/); // replay
  const ctr2 = unhex(KAT.frame);
  ctr2[7] = 3; // skip ahead
  await assert.rejects(sess.open(ctr2), /counter/);
  // tampered ciphertext with the right counter
  const t = unhex(KAT.frame);
  t[7] = 2;
  t[t.length - 1] ^= 1;
  await assert.rejects(sess.open(t), /authenticate/);
  // outbound counters 1, 2, 3 ...
  const a = await sess.seal(new Uint8Array([1]));
  const b = await sess.seal(new Uint8Array([2]));
  assert.equal(a[7], 1);
  assert.equal(b[7], 2);
});

test("deriveSession is deterministic and fingerprint formats as chalkctl prints it", async () => {
  const th = fill(9);
  const ss = fill(7);
  const s1 = await deriveSession(ss, th);
  const s2 = await deriveSession(ss, th);
  const f1 = await s1.seal(new Uint8Array([42]));
  const f2 = await s2.seal(new Uint8Array([42]));
  assert.equal(hex(f1), hex(f2));
  const fp = await serverFingerprint(fill(1));
  assert.match(fp, /^([0-9a-f]{4} ){7}[0-9a-f]{4}$/);
});
